sputnikdao-factory2/src/lib.rs

```diff
@@ -1,106 +1,77 @@
 
-impl Default for SputnikDAOFactory {
-    fn default() -> Self {
-        env::panic_str("SputnikDAOFactory should be initialized before usage")
-    }
-}
-
 #[near_bindgen]
 impl SputnikDAOFactory {
     #[init]
     pub fn new() -> Self {
-        assert!(!env::state_exists(), "The contract is already initialized");
         Self {
+            factory_manager: FactoryManager {},
             daos: UnorderedSet::new(b"d".to_vec()),
         }
     }
 
-    pub fn get_dao_list(&self) -> Vec<AccountId> {
-        self.daos.to_vec()
+    pub fn set_owner(&self, owner_id: AccountId) {
+        self.assert_owner();
+        env::storage_write(OWNER_KEY, owner_id.as_bytes());
     }
 
-    /// Get number of created DAOs.
-    pub fn get_number_daos(&self) -> u64 {
-        self.daos.len()
+    pub fn set_code_hash(&self, code_hash: Base58CryptoHash) {
+        self.assert_owner();
+        let code_hash: CryptoHash = code_hash.into();
+        env::storage_write(LATEST_CODE_HASH_KEY, &code_hash);
     }
 
-    /// Get daos in paginated view.
-    pub fn get_daos(&self, from_index: u64, limit: u64) -> Vec<AccountId> {
-        let elements = self.daos.as_vector();
-        (from_index..std::cmp::min(from_index + limit, elements.len()))
-            .filter_map(|index| elements.get(index))
-            .collect()
+    pub fn delete_contract(&self, code_hash: Base58CryptoHash) {
+        self.assert_owner();
+        self.factory_manager.delete_contract(code_hash);
     }
 
     #[payable]
-    pub fn create(
-        &mut self,
-        name: AccountId,
-        public_key: Option<PublicKey>,
-        args: Base64VecU8,
-    ) -> Promise {
+    pub fn create(&mut self, name: AccountId, args: Base64VecU8) {
         let account_id: AccountId = format!("{}.{}", name, env::current_account_id())
             .parse()
             .unwrap();
-        let mut promise = Promise::new(account_id.clone())
-            .create_account()
-            .deploy_contract(CODE.to_vec())
-            .transfer(env::attached_deposit());
-        if let Some(key) = public_key {
-            promise = promise.add_full_access_key(key.into())
-        }
-        promise
-            .function_call(
-                "new".to_string(),
-                args.into(),
-                0,
-                env::prepaid_gas() - CREATE_CALL_GAS - ON_CREATE_CALL_GAS,
-            )
-            .then(ext_self::on_create(
-                account_id,
-                U128(env::attached_deposit()),
-                env::predecessor_account_id(),
-                env::current_account_id(),
-                0,
-                ON_CREATE_CALL_GAS,
-            ))
+        let callback_args = serde_json::to_vec(&json!({
+            "account_id": account_id,
+            "attached_deposit": U128(env::attached_deposit()),
+            "predecessor_account_id": env::predecessor_account_id()
+        }))
+        .expect("Failed to serialize");
+        self.factory_manager.create_contract(
```

Comment:
Compared with `update_contract`, this looks good since we have the `new` method inside the DAO.

Reply:
ignore this comment as well